internal/campaigns/execution_cache.go

```diff
@@ -24,6 +24,33 @@ type ExecutionCacheKey struct {
 	*Task
 }
 
+func (key ExecutionCacheKey) MarshalJSON() ([]byte, error) {
+	// We have to resolve the step environments and include them in the cache
+	// key to ensure that the cache is properly invalidated when an environment
+	// variable changes.
+	//
+	// Note that we don't base the cache key on the entire global environment:
+	// if an unrelated environment variable changes, that's fine. We're only
+	// interested in the ones that actually make it into the step container.
+	global := os.Environ()
+	envs := make([]map[string]string, len(key.Task.Steps))
+	for i, step := range key.Task.Steps {
+		env, err := step.Env.Resolve(global)
+		if err != nil {
+			return nil, errors.Wrapf(err, "resolving environment for step %d", i)
+		}
+		envs[i] = env
+	}
+
+	return json.Marshal(struct {
+		*Task
+		Environments []map[string]string
+	}{
+		Task:         key.Task,
+		Environments: envs,
+	})
+}
```

Comment:
It took me a while to realise that we don't persist the the plain text version of this on disk, but take a `sha256` sum and the `base64` of it in `(ExecutionDiskCache).cacheFilePath`.

And now, with more logic in `MarshalJSON` maybe we should make the behaviour more explicit and just give `ExecutionCacheKey` a `String()` method that contains all of it — the json marshalling, the sha sum, the base64 encoding — instead of the logic being split in two methods?

Reply:
Yeah, I had a similar thought. I think I'll go ahead and do that in this PR, since I agree it makes sense.